### ISSUE TITLE
Runtime Error Question

### ISSUE THREAD

**Issue by darebear** [christianrauch/multiwii_ros2#2]
Hello again,

I obtained the following when setting the "port" in Betaflight to Configuration/MSP: 

```
dario@DESKTOP-S9T430H:~/ros2_example_ws$ ros2 run multiwii_node  multiwii_node 
open /dev/ttyUSB0@115200
Message marker 83 is not recognised!
Version marker 65 is not recognised!
terminate called after throwing an instance of 'std::runtime_error'
  what():  Cannot get BoxNames!
``` 

and obtain this when setting the port to "Serial RX":

```
dario@DESKTOP-S9T430H:~/ros2_example_ws$ ros2 run multiwii_node  multiwii_node 
open /dev/ttyUSB0@115200
```

When I run the examples from https://github.com/christianrauch/msp, I will sometimes obtain this message:

```
dario@DESKTOP-S9T430H:~/msp$ ./build/client_async_test 
... 
Print method for message ID 111 is not implemented
#Altitude:
 Altitude: 0 m, var: 0 m/s
 Barometer: <unset>
Print method for message ID 112 is not implemented
Message marker 83 is not recognised!
Version marker 65 is not recognised!
Message marker 30 is not recognised!
Version marker 31 is not recognised!
Message marker 0 is not recognised!
Version marker 0 is not recognised!
Message marker 0 is not recognised!
Version marker 229 is not recognised!
#Imu:
 Linear acceleration: 162, 29, 486
...
```

when I run the following, I get a similar error to the first one posted above: 

```
dario@DESKTOP-S9T430H:~/msp$ ./build/client_read_test 
Message v1 with ID 100 is not recognised!
unsupported: 100
#Status:
 Cycle time: 312 us
 I2C errors: 0
 Sensors:
    Accelerometer: ON
    Barometer: OFF
    Magnetometer: OFF
    GPS: OFF
    Sonar: OFF
 Active Boxes (by ID):
#Imu:
 Linear acceleration: 165, 30, 484
 Angular velocity: 1, -5, 1
 Magnetometer: 0, 0, 0
#Imu:
 Linear acceleration: 3.16035, 0.574608, 9.27035 m/s²
 Angular velocity: 0.244141, -1.2207, 0.244141 deg/s
 Magnetometer: 0, 0, 0 uT
#Servo:
 1500 1500 1500 1500
 1500 1500 1500 1500
#Motor:
 1000 1000 1000 1000
 0 0 0 0
#Rc channels (18) :
1500 1500 1500 885 1425 1500 1500 1500 1500 1500 1500 1500 1500 1500 1500 1500 1500 1500  
#Attitude:
 Roll : 3.4 deg
 Pitch : -18.6 deg
 Heading: 0 deg
#Altitude:
 Altitude: 0 m, var: 0 m/s
 Barometer: <unset>
#Analog:
 Battery Voltage: 5.2 V
 Current: 0 A
 Power consumption: 0 Ah
 RSSI: 0
Print method for message ID 111 is not implemented
unsupported: 112
Message v1 with ID 113 is not recognised!
unsupported: 113
Message v1 with ID 114 is not recognised!
unsupported: 114
Message v1 with ID 115 is not recognised!
unsupported: 115
Message v1 with ID 116 has wrong CRC! (expected: 128, received: 73)
Message marker 83 is not recognised!
Version marker 65 is not recognised!
buffer returned EOF; reading char directly from port
unsupported: 116
buffer returned EOF; reading char directly from port
...
```

Is the expected behavior that upon successful connection to the fcu via `fcu->connect(device_path, baud);` that the node will start publishing the different topics? I ask because I've put `std::cout << "fcu connected" << std::endl;` after `fcu->connect(device_path, baud);` and I am not seeing that message. 

Thank you.

**Comment by christianrauch** [comment]
You will get messages like `Message v1 with ID <ID> is not recognised!` if a message type is not supported by the firmware. The supported message types differ between firmware as different forks sometimes use a different set if message types.

The messages like:
```
Message marker <value> is not recognised!
Version marker <value> is not recognised!
```
mean that there was an error while parsing the byte stream and searching for the "magic bytes". This could be a communication issue on the serial connection.

> Is the expected behavior that upon successful connection to the fcu via fcu->connect(device_path, baud); that the node will start publishing the different topics?

What do you mean by "different topics"? The ROS topics which the node uses are defined in the code. They do not change dynamically.

> and I am not seeing that message

It might simply be that this method has not returned yet. The `connect` method sends a couple of messages to the FCU and waits for responses. If those responses do not arrive, or the message isn't supported at all, then this might block forever.

I haven't tested recent betaflight versions since they deprecated the hardware that I am using. It might be that they broke backwards compatibility by removing or changing message definitions.

**Comment by darebear** [comment]
> What do you mean by "different topics"? The ROS topics which the node uses are defined in the code. They do not change dynamically.

Ok, this makes a lot more sense now. Since the node was not publishing any messages, I assumed one had to call on them by instantiating the `MultiWiiNode` class (as I mentioned previously). I concur that the `connect` method might be blocking forever. I will look deeper into this issue as I want to use `msp` to communicate with BF boards. 

Thank you.

**Comment by darebear** [comment]
> I haven't tested recent betaflight versions since they deprecated the hardware that I am using. It might be that they broke backwards compatibility by removing or changing message definitions.

I rolled my board to the earliest firmware version available, `4.1.0 Legacy - 16-10-2019 07:29`, and now `fcu->connect` is returning, and the rest of the node seems to be working as expected since the topics are published. Thank you again.
